src/Build.OM.UnitTests/Construction/ProjectFormatting_Tests.cs

```diff
@@ -605,5 +603,86 @@ public void ChangeItemTypeWithXmlHeader()
+
+            VerifyAssertLineByLine(expected, actual);
+        }
+
+        [Fact]
+        public void ProjectFileWithBomContainsBomAfterSave()
+        {
+            string content = ObjectModelHelpers.CleanupFileContents(@"<Project />");
+
+            var file = FileUtilities.GetTemporaryFile(".proj");
+            try
+            {
+                File.WriteAllText(file, content, new UTF8Encoding(true));
+                File.SetLastWriteTime(file, DateTime.Now.Subtract(TimeSpan.FromDays(1)));
+                Assert.True(FileStartsWithBom(file));
+
+                // Load and manipulate/save the project
+                var project = new Project(ProjectRootElement.Open(file, ProjectCollection.GlobalProjectCollection));
+                project.AddItem("Compile", "Program.cs");
+                project.Save();
+
+                // Ensure the file was really saved and that the file still contains a BOM
+                Assert.True(File.GetLastWriteTime(file).Subtract(DateTime.Now).TotalSeconds < 2);
+                Assert.True(File.ReadAllText(file).Contains("<Compile Include=\"Program.cs\" />"));
+                Assert.True(FileStartsWithBom(file));
+            }
+            finally
+            {
+                FileUtilities.DeleteDirectoryNoThrow(Path.GetDirectoryName(file), false);
+            }
+        }
+
+        [Fact]
+        public void ProjectFileWithoutBomDoesNotContainsBomAfterSave()
+        {
+            string content = ObjectModelHelpers.CleanupFileContents(@"<Project><Target Name=""Build""/></Project>");
+
+            var file = FileUtilities.GetTemporaryFile(".proj");
+            try
+            {
+                File.WriteAllText(file, content, new UTF8Encoding(false));
+                File.SetLastWriteTime(file, DateTime.Now.Subtract(TimeSpan.FromDays(1)));
+                Assert.False(FileStartsWithBom(file));
+
+                // Load and manipulate/save the project
+                var project = new Project(ProjectRootElement.Open(file, ProjectCollection.GlobalProjectCollection));
+                project.AddItem("Compile", "Program.cs");
+                project.Save();
+
+                Assert.False(FileStartsWithBom(file));
+                Assert.True(File.GetLastWriteTime(file).Subtract(DateTime.Now).TotalSeconds < 2);
+                Assert.True(File.ReadAllText(file).Contains("<Compile Include=\"Program.cs\" />"));
+            }
+            finally
+            {
+                FileUtilities.DeleteDirectoryNoThrow(Path.GetDirectoryName(file), false);
+            }
+        }
+
+        private static bool FileStartsWithBom(string file)
+        {
+            using (var stream = new FileStream(file, FileMode.Open, FileAccess.Read, FileShare.Read))
+            {
+                var bom = new byte[3];
+                stream.Read(bom, 0, 3);
+                var preamble = new UTF8Encoding(true).GetPreamble();
+                return !preamble.Where((p, i) => p != bom[i]).Any();
```

Comment:
Normally I say to LINQify things but I think this would be clearer as a for loop :)